R/merge_sce_list.R

```diff
@@ -263,3 +302,131 @@ prepare_sce_for_merge <- function(sce,
+
+    # First, establish new rowData with all NA values
+    # this data.frame only contains rows for missing features
+    new_rowdata <- data.frame(
+      var = names(rowData(sce_altexp)),
+      val = rep(NA, n_missing)
+    ) |>
+      tidyr::pivot_wider(
+        names_from = var,
+        values_from = val
+      ) |>
+      # repeat rows n_missing times
+      dplyr::slice(
+        rep(
+          1:dplyr::n(),
+          each = n_missing
+        )
+      )
+
+    # create new rownames to be added in when we re-DataFrame this
+    new_rownames <- c(
+      rownames(sce_altexp), # existing rownames
+      missing_features # new rownames
+    )
+
+    # Next, establish new counts and logcounts assays
+    # TODO: Can either add `if` for whether to run on each assay, or fail if
+    #  either is missing. Any reviewer preferences?
+    new_counts <- update_altexp_assay(
+      counts(sce_altexp),
+      missing_features
+    )
+
+    new_logcounts <- update_altexp_assay(
+      logcounts(sce_altexp),
+      missing_features
+    )
+
+    # Create a new altexp and add it into the SCE
+    # TODO: this line needs to be changed if we want to do `if` with assay update
+    new_altexp <- SingleCellExperiment(assays = list(counts = new_counts,
+                                                     logcounts = new_logcounts))
+
+    # Add the new rowData rows into the SCE, with updated column names
+    rowData(new_altexp) <- sce_altexp |>
+      rowData() |>
+      as.data.frame() |>
+      rbind(new_rowdata) |>
+      # re DataFrame with _all_ feature rownames
+      DataFrame(row.names = new_rownames)
+
+    # Add colData and metadata
+    colData(new_altexp) <- colData(altExp(sce))
+    metadata(new_altexp) <- metadata(altExp(sce))
+
+    # Replace the old altExp, while preparing it for merge
+    altExp(sce) <- prepare_sce_for_merge(
+      new_altexp,
+      sce_name,
+      batch_column = "library_id",
+      cell_id_column = "cell_id",
+      shared_features = altexp_features,
+      retain_coldata_cols = colnames(colData(new_altexp)),
+      preserve_rowdata_cols = NA # rename them all
+    )
+  }
+
+  # Return the SCE with updated altExp
+  return(sce)
+}
+
+
+
+#' Create a new assay sparse matrix with `NA` values for missing features
+#'
+#' @param assay_sparse_matrix Assay to modify, in sparse format
+#' @param missing_features Vector of missing features to add to the assay matrix
+#'
+#' @return Updated sparse matrix
+update_altexp_assay <- function(
+    assay_sparse_matrix,
+    missing_features) {
+
+  # First, create matrix for missing features with all NA values
+  new_matrix <- matrix(
+    NA,
+    nrow = length(missing_features),
+    ncol = ncol(assay_sparse_matrix)
+  )
+  rownames(new_matrix) <- missing_features
+
+  # Now rbind it to the existing matrix
+  new_assay <- assay_sparse_matrix |>
+    as.matrix() |>
+    rbind(new_matrix) |>
+    # make it sparse again
+    as("CsparseMatrix")
+
+  return(new_assay)
```

Comment:
I think we want to do this a bit differently, to be sure all of our matrices have the same row order, and it may simplify things too.

I think the below should work pretty directly if you pass in the full feature list.

```
# define the new full matrix
new_matrix <- matrix(nrow = length(all_features), ncol = ncol(assay_sparse_matrix), dimnames = list(all_features, rownames(assay_sparse_matrix))
# fill in existing features
new_matrix[rownames(assay_sparse_matrix),] <- matrix(assay_sparse_matrix)
```